src/aspire/abinitio/commonline_c3_c4.py

```diff
@@ -231,21 +237,270 @@ def _estimate_inplane_rotations(
+        n_img = self.n_img
+        n_theta = self.n_theta
+
+        nchoose2 = int(n_img * (n_img - 1) / 2)
+        Rijs = np.zeros((nchoose2, 3, 3))
+        pairs = all_pairs(n_img)
+        for idx, (i, j) in enumerate(pairs):
+            Rijs[idx] = self._syncmatrix_ij_vote_3n(
+                clmatrix, i, j, np.arange(n_img), n_theta
+            )
+
+        return Rijs
+
+    def _syncmatrix_ij_vote_3n(self, clmatrix, i, j, k_list, n_theta):
+        """
+        Compute the (i,j) rotation block of the synchronization matrix using voting method
+
+        Given the common lines matrix `clmatrix`, a list of images specified in k_list
+        and the number of common lines n_theta, find the (i, j) rotation block Rij.
+        :param clmatrix: The common lines matrix
+        :param i: The i image
+        :param j: The j image
+        :param k_list: The list of images for the third image for voting algorithm
+        :param n_theta: The number of points in the theta direction (common lines)
+        :return: The (i,j) rotation block of the synchronization matrix
+        """
+        good_k = self._vote_ij(clmatrix, n_theta, i, j, k_list)
+
+        rots = self._rotratio_eulerangle_vec(clmatrix, i, j, good_k, n_theta)
+
+        if rots is not None:
+            rot_mean = np.mean(rots, 0)
+
+        else:
+            # This is for the case that images i and j correspond to the same
+            # viewing direction and differ only by in-plane rotation.
+            # We set to zero as in the Matlab code.
+            rot_mean = np.zeros((3, 3))
+
+        return rot_mean
+
+    def _local_J_sync_c3_c4(self, Rijs, Riis):
+        """
+        Estimate viis and vijs. In order to estimate vij = vi @ vj.T, it is necessary for Rii, Rjj,
+        and Rij to be of the same handedness. We perform a local handedness synchronization and
+        set vij = 1/n*sum(Rii^s @ Rij @ Rjj^s).
+
+        :param Rijs: An n-choose-2x3x3 array of estimates of relative rotations
+            (each pair of images induces two estimates).
+        :param Riis: A nx3x3 array of estimates of self-relative rotations.
+        :return: vijs, viis
+        """
+        order = self.order
+        n_img = self.n_img
+
+        nchoose2 = int(n_img * (n_img - 1) / 2)
+        assert (
+            len(Riis) == n_img
+        ), f"There must be one self-relative rotation per image. Got {len(Riis)} Riis."
+        assert (
+            len(Rijs) == nchoose2
+        ), f"There must be n-choose-2 relative rotations. Got {len(Rijs)}."
+
+        # Estimate viis from Riis. vii = 1/order * sum(Rii^s) for s = 0, 1, ..., order.
+        viis = np.zeros((n_img, 3, 3))
+        for i, Rii in enumerate(Riis):
+            viis[i] = np.mean(
+                [np.linalg.matrix_power(Rii, s) for s in np.arange(order)], axis=0
+            )
+
+        # Estimate vijs via local handedness synchronization.
+        vijs = np.zeros((nchoose2, 3, 3))
+        e1 = [1, 0, 0]
```

Comment:
Define these closer to where they are used.

Reply:
Done.